File: algorithms/src/test/java/jkt/patterns/algorithms/BagTest.java
package jkt.patterns.algorithms;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Iterator;

public class BagTest {
  Bag<String> bag;

  private void setUpStack() {
    bag = new Bag<>();
    bag.add("lions");
    bag.add("tigers");
    bag.add("bears");
  }

  public BagTest() {
    this.setUpStack();
  }

  @Test
  public void bagWithItemsIsNotEmpty() {
    Assertions.assertEquals(false, bag.isEmpty());
  }

  @Test
  public void bagHasExpectedSize() {
    Assertions.assertEquals(3, bag.size());
  }

  @Test
  public void canIterateOverItems() {
    Iterator<String> iterator = bag.iterator();
    Assertions.assertEquals(true, iterator.hasNext());
    Assertions.assertEquals("bears", iterator.next());
    Assertions.assertEquals("tigers", iterator.next());
    Assertions.assertEquals("lions", iterator.next());
  }
}

File: algorithms/src/test/kotlin/jkt/patterns/algorithms/BagTest.kt
package jkt.patterns.algorithms

import org.junit.jupiter.api.Test

import org.junit.jupiter.api.Assertions.*

internal class BagTest {

  @Test
  fun isEmpty() {
    val bag = Bag<Int>()
    assertEquals(true, bag.isEmpty)
    bag.add(11)
    assertEquals(false, bag.isEmpty)
  }

  @Test
  fun size() {
    val bag = Bag<Int>()
    assertEquals(0, bag.size())
    assertEquals(3, bag.add(11).add(13).add(17).size())
  }

  @Test
  fun add() {
    val bag = Bag<String>()
    assertEquals(0, bag.size())
    assertEquals(3, bag.add("aaa").add("bbb").add("ccc").size())
  }

  @Test
  operator fun iterator() {
    val bag = Bag<Int>().add(3).add(5).add(7).add(11).add(13).add(17).add(19)
    val expected = listOf(3, 5, 7, 11, 13, 17, 19)
    val result = mutableListOf<Int>()
    bag.forEach { result.add(it) }
    assertEquals(expected.sorted(), result.sorted())
  }
}